middle_end/flambda2/algorithms/patricia_tree.ml

```diff
@@ -406,9 +453,13 @@ end = struct
     | Leaf (i, d0), Leaf (j, d1) when i = j -> (
       (* CR mshinwell: [join] in [Typing_env_level] is relying on the fact that
          the arguments to [f] are always in the correct order, i.e. that the
-         first datum comes from [t0] and the second from [t1]. Document. *)
+         first datum comes from [t0] and the second from [t1]. Document.
+
+         lmaurer: This is a good reason to implement [union] in terms of (a
+         generalized) [merge], since implementing [merge] requires the types to
+         line up like this. *)
       match f i d0 d1 with None -> empty iv | Some datum -> leaf iv i datum)
```

Comment:
It seems that `join` doesn't have this constraint any more (it doesn't use `union` at all, by the look of it) -- so this CR can be removed.  Maybe replace it with a CR-someday with the suggestion about implementing `union` in terms of `merge`.